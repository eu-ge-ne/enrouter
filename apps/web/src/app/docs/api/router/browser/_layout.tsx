import Index from "./index.md";

export const components = {
  api: Layout,
};

function Layout() {
  return <Index />;
}
