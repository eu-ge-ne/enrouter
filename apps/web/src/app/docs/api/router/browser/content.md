# BrowserRouter API

## Components

### BrowserRouter

```typescript
interface BrowserRouterProps {
  handlers: RouteHandler;
  modules: RouteModules;
}

function BrowserRouter({ handlers, modules }: BrowserRouterProps);
```
