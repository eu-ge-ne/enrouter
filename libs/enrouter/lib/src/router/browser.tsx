import { type ReactNode, useState, useCallback, useEffect } from "react";

import type { Route } from "#lib/route/mod.js";
import type { Match } from "#lib/match/mod.js";
import { logger } from "#lib/debug.js";
import { matchRoutes } from "#lib/match/match.js";
import { loadMatches } from "#lib/match/load.js";
import { renderMatches } from "#lib/match/render.js";
import {
  type TRouterDynamicContext,
  RouterStaticProvider,
  RouterDynamicProvider,
} from "./context.js";

const log = logger("router/browser");

export interface BrowserRouterProps {
  routes: Route;
  matches: Match[];
}

export function BrowserRouter({
  routes,
  matches: initialMatches,
}: BrowserRouterProps): ReactNode {
  const [dynamicContext, setDynamicContext] = useState<TRouterDynamicContext>({
    location: window.location.pathname,
    matches: initialMatches,
    children: renderMatches(initialMatches),
  });

  const navigate = useCallback(async (location: string) => {
    log("Navigating to %s", location);

    window.history.pushState({}, "", location);

    const matches = matchRoutes({ routes, location });
    await loadMatches(matches);
    const children = renderMatches(matches);

    setDynamicContext({ location, matches, children });
  }, []);

  const handlePopState = useCallback(async (e: PopStateEvent) => {
    log("handlePopState %o", e);

    const location = window.location.pathname;
    const matches = matchRoutes({ routes, location });
    await loadMatches(matches);
    const children = renderMatches(matches);

    setDynamicContext({ location, matches, children });
  }, []);

  useEffect(() => {
    window.addEventListener("popstate", handlePopState);
    return () => window.removeEventListener("popstate", handlePopState);
  }, [handlePopState]);

  const staticContext = { routes, navigate };

  return (
    <RouterStaticProvider value={staticContext}>
      <RouterDynamicProvider value={dynamicContext}>
        {dynamicContext.children}
      </RouterDynamicProvider>
    </RouterStaticProvider>
  );
}
