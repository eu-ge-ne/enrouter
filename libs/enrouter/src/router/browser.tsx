import { useState, useCallback, useEffect, useMemo } from "react";

import { createLog } from "#log.js";
import { matchRoutes } from "#match/mod.js";
import { loadRouteMatches } from "#loader/match.js";
import { renderMatches } from "#render/mod.js";
import { RouterContext } from "./context.js";

import type { RouteHandler } from "#handler/mod.js";
import type { RouteModules } from "#modules.js";
import type { TRouterContext } from "./context.js";

const log = createLog("router/browser");

interface BrowserRouterProps {
  handlers: RouteHandler;
  modules: RouteModules;
}

export function BrowserRouter({ handlers, modules }: BrowserRouterProps) {
  const [location, setLocation] = useState(window.location.pathname);

  const handlePopState = useCallback((e: PopStateEvent) => {
    log("handlePopState %O", e);
    setLocation(window.location.pathname);
  }, []);

  useEffect(() => {
    window.addEventListener("popstate", handlePopState);
    return () => window.removeEventListener("popstate", handlePopState);
  }, [handlePopState]);

  const navigate = useCallback(async (to: string) => {
    log("Navigating to %O", to);

    const matches = matchRoutes({ handlers, location: to });
    await loadRouteMatches({ matches, modules });

    window.history.pushState({}, "", to);

    setLocation(to);
  }, []);

  const matches = useMemo(
    () => matchRoutes({ handlers, location }),
    [handlers, location],
  );

  const context = useMemo<TRouterContext>(
    () => ({
      handlers,
      location,
      navigate,
    }),
    [handlers, location, navigate],
  );

  const children = useMemo(() => renderMatches(matches), [matches]);

  return (
    <RouterContext.Provider value={context}>{children}</RouterContext.Provider>
  );
}
