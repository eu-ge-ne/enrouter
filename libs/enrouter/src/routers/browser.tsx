import { useState, useCallback, useEffect, useMemo } from "react";

import { createLog } from "#log.js";
import { matchRoutes } from "#matches/mod.js";
import { loadRouteMatches } from "#load/matches.js";
import { renderMatches } from "#render/mod.js";
import { RouterContext } from "./context.js";

import type { RouteHandler } from "#handlers/mod.js";
import type { RouteModules } from "#modules.js";
import type { TRouterContext } from "./context.js";

const log = createLog("routers/browser");

interface BrowserRouterParams {
  handlers: RouteHandler;
  modules: RouteModules;
}

export function BrowserRouter({ handlers, modules }: BrowserRouterParams) {
  const [location, setLocation] = useState(window.location.pathname);

  const handlePopState = useCallback((e: PopStateEvent) => {
    log("handlePopState %O", e);
    setLocation(window.location.pathname);
  }, []);

  useEffect(() => {
    window.addEventListener("popstate", handlePopState);
    return () => window.removeEventListener("popstate", handlePopState);
  }, [handlePopState]);

  const navigate = useCallback(async (to: string) => {
    log("Navigating to %O", to);

    const matches = matchRoutes({ handlers, location: to });
    await loadRouteMatches({ matches, modules });

    window.history.pushState({}, "", to);

    setLocation(to);
  }, []);

  const matches = useMemo(
    () => matchRoutes({ handlers, location }),
    [handlers, location],
  );

  const context = useMemo<TRouterContext>(
    () => ({
      handlers,
      location,
      navigate,
    }),
    [handlers, location, navigate],
  );

  const children = useMemo(() => renderMatches(matches), [matches]);

  return (
    <RouterContext.Provider value={context}>{children}</RouterContext.Provider>
  );
}
