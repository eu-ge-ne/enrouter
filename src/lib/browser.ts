export function pushHistory(to: string) {
  window.history.pushState({}, "", to);
}

export function assignLocation(to: string) {
  window.location.assign(to);
}
