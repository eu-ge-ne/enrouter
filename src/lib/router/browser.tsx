import {
  type ReactNode,
  type ComponentType,
  useState,
  useCallback,
  useEffect,
} from "react";

import { logger } from "#lib/debug.js";
import { type Match, matchLocation } from "#lib/match/match.js";
import * as browser from "#lib/browser/mod.js";
import { Root } from "#lib/root/root.js";

const log = logger("router/browser");

export interface BrowserRouterProps {
  root?: ComponentType;
  void?: Record<string, ComponentType>;
  matches: Match[];
}

export function BrowserRouter(props: BrowserRouterProps): ReactNode {
  const [location, setLocation] = useState(window.location.pathname);
  const [matches, setMatches] = useState(props.matches);

  const navigate = useCallback(async (to: string) => {
    try {
      const newMatches = await matchLocation(to);

      setLocation(to);
      setMatches(newMatches);

      browser.pushHistory(to);

      log("Navigated to %s", to);
    } catch (err) {
      log("matchLocation error %o", err);

      browser.assignLocation(to);
    }
  }, []);

  const handlePopState = useCallback(async (e: PopStateEvent) => {
    log("handlePopState %o", e);

    const to = window.location.pathname;
    setLocation(to);

    setMatches(await matchLocation(to));
  }, []);

  useEffect(() => {
    window.addEventListener("popstate", handlePopState);
    return () => window.removeEventListener("popstate", handlePopState);
  }, [handlePopState]);

  return (
    <Root
      params={{ root: props.root, void: props.void }}
      navigate={navigate}
      location={location}
      matches={matches}
    />
  );
}
