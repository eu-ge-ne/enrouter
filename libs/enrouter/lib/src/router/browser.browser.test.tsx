import type { FC, PropsWithChildren } from "react";
import { describe, test, expect } from "vitest";
import { render } from "vitest-browser-react";
import * as regexparam from "regexparam";

import type { Route } from "#lib/route/mod.js";
import type { Match } from "#lib/match/mod.js";
import { Browser } from "./browser.js";

const wrapperId = "test-wrapper";

const wrapper: FC<PropsWithChildren> = ({ children }) => (
  <div data-testid={wrapperId}>{children}</div>
);

describe("router", () => {
  describe("Browser", () => {
    test("_root", async () => {
      const route: Route = {
        path: "/",
        test: regexparam.parse("/", true),
        modules: [],
        loaded: true,
        elements: {
          _root: <div>Root</div>,
        },
      };

      const match: Match = {
        route,
        location: "/",
        isFull: true,
        params: {},
      };

      const screen = render(<Browser routes={route} match={match} />, {
        wrapper,
      });

      await expect.element(screen.getByTestId(wrapperId)).toBeVisible();

      expect(screen.container).toMatchSnapshot();
    });
  });
});
