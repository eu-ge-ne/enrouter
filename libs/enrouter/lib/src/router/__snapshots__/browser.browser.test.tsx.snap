// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`router > Browser > _root 1`] = `
<div>
  <div
    data-testid="test-wrapper"
  >
    <div>
      Root
    </div>
  </div>
</div>
`;
